### ISSUE TITLE
Support workspace symbol

### ISSUE THREAD

**Issue by matklad** [zigtools/zls#778]
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_symbol

The TL;DR of the feature is that the user inputs a string, and the server provides a fuzzy-matched list of functions, types, etc, which match the query. 

This is another relatively-easy-to-implement/high-impact feature (this is the main way to navigate large codebases generally).

Implementation-wise, it needs three things:

* ability to parse the code and extract named definitions. Zig is going to be a bit tricky, as structs are anonymous, so we need some heuristic to handle things like
    ```
        pub const Options = struct {
            configuration: []const std.net.Address,
            io: *IO,
        };
    ```

    Though presumably just indexing everything starting with a capital letter would work
* some sort of map-reduce-like infra to "index" all source-file, and update the index when file are edit
* a fuzzy search which can filter the total index with the user's query

**Comment by matklad** [comment]
Notes: 

Existing "documentSymbols" contains a fair bit of machinery we'll need here: 

https://github.com/zigtools/zls/blob/1ced17266c0f6256c7c0ea515e2407ff56c2ee6b/src/Server.zig#L2083-L2092

In a nutshell, "workspaceSymbol" is "union of `documentSymbol` for all documents + a fuzzy index". 

What is unclear at this point is how to iterate all documents in the workspace. Looking at the DocumentStore

https://github.com/zigtools/zls/blob/1ced17266c0f6256c7c0ea515e2407ff56c2ee6b/src/DocumentStore.zig#L89-L90

it looks like it doesn't support this yet. More generally, it uses a different strategy than we need here: 

* DocumentStore keeps a set of open files and a transitive set of their depenrencies
* for workspace symbols, we want to be aware about all existing files, reagrdless of whether they are opened or not.  

That's a pretty core thing: compilers generally like to _lazily_ explore tree of files from the root, while IDEs like to be _eagerly_ aware of everything under the sun. 

As an MVP, we can just build an index of all the symbols document store is aware of of course :)

**Comment by Techatrix** [comment]
duplicate of #397
